### ISSUE TITLE
Deploying an application for any hardware

### ISSUE THREAD

**Issue by j8asic** [kokkos/kokkos#1432]
What would be the proper way to:
- deploy an application that can cover all possible execution spaces (nvidia, amd, openmp, ...)
- choose the best device in the system
Best regards

**Comment by mhoemmen** [comment]
That (esp. the second question) sounds like many PhD dissertations worth of research.  What's "best"?  It depends on what algorithms you're running and what problems you're solving.

Kokkos gives you the tools to answer your first question.  However, it doesn't solve the problem for you.  You still have to decide where (in what execution space) you want to run, and whither (to what memory space) you want to move data.

I would expect applications or libraries to build infrastructure on top of Kokkos that coordinates asynchronous kernel launches and data movement.

**Comment by crtrott** [comment]
I am not sure that you could do this if by deploy you mean distribute a single binary. I am not sure that you can actually link all the stuff together. I somewhat fear you may need to provide multiple binaries. 

In contrast to the graphics market there is no good programming language flavor which actually works well across all architectures and still provides C++ support.

It might (and I mean might) be possible to do something like have a wrapper library which is fully templated on execution space and you instantiate that depending on what you detect, potentially even via "dlopen" loading only the specific library, i.e. something like this:

```c++
int main() {
  if(DeviceType() = NVIDIA)
    my_app<Cuda>();
  if(DeviceType() = AMD)
    my_app<ROCm>();
}
```

Or:
```c++

int main() {
  if(DeviceType() = NVIDIA)
   //dlopen("libapp_nvidia.so");
  if(DeviceType() = AMD)
   //dlopen("libapp_amd.so");
}
```

Or something in that direction. Note that Kokkos's build system and some of its internals don't even support fat binaries within one ecosystem right now. For example you can't compiler for two different NVIDIA architectures at once. While there is no strong technical reason why this wouldn't be possible there are some places in Kokkos where it would mean some work to enable that deployment method. But since we are focused on HPC we had never a driver to do that. Basically in HPC most folks recompile their apps for every machine anyway. 

What would your requirements be, and can you tell us something about your driver to do so?

**Comment by mhoemmen** [comment]
Kokkos does support multiple execution spaces, but you could only support one version of each.  (For example, you can have CUDA and OpenMP, but not CUDA 8.0 and CUDA 9.0, and not "OpenMP with GCC and OpenMP with Intel.")

**Comment by j8asic** [comment]
Thanks for the answers!

At the moment I actually have an application that is templated for each execution space (also templated based on the floating point precision, problem dimensions, etc.). It's not a problem for me to use specific parameters at runtime.

What I never dealt with, is how to have it in a single binary. With the rise of AMD GPUs in scientific area, I would like to offer a single binary that would choose the "best" solver within the application. An example: if nVidia gpu available then use it, else if AMD gpu is available use it, else use OpenMP.
So basically I like your first code example. The concerns are: linking, deploying files, and ofc testing for available execution spaces in the code.

**Comment by mhoemmen** [comment]
@j8asic AMD's GPU uses a different Kokkos back-end: 
https://github.com/kokkos/kokkos/blob/master/core/src/Kokkos_ROCm.hpp

Christian points out the issue that you might need to support different versions of the same programming models -- e.g., CUDA 8 vs. CUDA 9 -- for different hardware.  I would suggest isolating the platform-specific stuff in separate dynamic libraries, then loading those in (e.g., `dlopen` or equivalent) after detecting the user's hardware.  Compare to how Matlab stores different BLAS library versions and pulls in the right one at run time (I think you can even control this via environment variable).

**Comment by crtrott** [comment]
I guess this answered your question I am closing the issue now. If not feel free to reopen.
